[Batch] Patch TaskOperations.add_collection with convenience functionality

bgklein opened the pull request:
Patch bulk task add functionality from Batch Extensions SDK to the base sdk.

lmazuel commented on azure-batch/azure/batch/custom/patch.py:
```diff
@@ -0,0 +1,285 @@
+                if len(chunk_tasks_to_add) == 1:
+                    failed_task = chunk_tasks_to_add.pop()
+                    results_queue.appendleft(e)
+                    logging.error("Task ID " + failed_task.id + " failed to add due to exceeding the request body" +
```
Never use logging that.
At the beginning of the file, after the import clauses, do a:
```python
_LOGGER = logging.getLogger(__name__)
```
and then use `_LOGGER`

lmazuel commented on azure-batch/azure/batch/custom/patch.py:
```diff
@@ -0,0 +1,285 @@
+
+def patch_client(client):
+    batch_client = client
+    def callback_bulk_add_collection(*args, **kwargs):
```
Use wraps, otherwise what the IDE sees is the callback, and you loose all intellisense and doc:
https://docs.python.org/3/library/functools.html#functools.wraps

lmazuel replied:
You can also uses types to avoid the callback
```python
target.method = types.MethodType(method,target)
```
which use be for you:
```python
client.task.add_collection = types.MethodType(bulk_add_collection, client.task)
```

This would be the prefered approach anyway, since you will loose the `metadata` dict otherwise (won't be copied with the callback)

lmazuel commented on azure-batch/azure/batch/version.py:
```diff
@@ -9,5 +9,5 @@
 # --------------------------------------------------------------------------
 
-VERSION = "5.0.0"
+VERSION = "5.0.1"
```
Would be 5.1.0 to respect semver (new features)

lmazuel commented on azure-batch/azure/batch/custom/patch.py:
```diff
@@ -0,0 +1,285 @@
+    :rtype: list[~TaskAddResult]
+    """
+    results = []
+    while len(results_queue) != 0:
```
```python
while results_queue:
```

lmazuel commented on azure-batch/azure/batch/custom/patch.py:
```diff
@@ -0,0 +1,285 @@
+    while len(results_queue) != 0:
+        queue_item = results_queue.pop()
+        if isinstance(queue_item, Exception):
+            raise queue_item
```
Sounds a bad idea, you loose the initial stacktrace by doing that. There should be a way to re-design this.

lmazuel commented on azure-batch/azure/batch/custom/patch.py:
```diff
@@ -0,0 +1,285 @@
+from . import errors
+from .. import models
+from ..models import BatchErrorException, TaskAddCollectionResult, TaskAddCollectionParameter
+from ..models.batch_service_client_enums import TaskAddStatus
```
merge this into the previous line
```python
from ..models import BatchErrorException, TaskAddCollectionResult, TaskAddCollectionParameter, TaskAddStatus
```

lmazuel commented on azure-batch/azure/batch/custom/patch.py:
```diff
@@ -0,0 +1,285 @@
+                    with self._max_tasks_lock:
+                        if midpoint < self._max_tasks_per_request:
+                            self._max_tasks_per_request = midpoint
+                            logging.info("Amount of tasks per request reduced from " +
```
Never log with "+" concat. Use the comma syntax, otherwise your evaluate your string even if the log level is undefined:
https://reinout.vanrees.org/weblog/2015/06/05/logging-formatting.html

lmazuel commented on azure-batch/azure/batch/custom/patch.py:
```diff
@@ -0,0 +1,285 @@
+        except Exception as e:  # pylint: disable=broad-except
+            results_queue.appendleft(e)
+        else:
+            if isinstance(add_collection_response, ClientRawResponse):
```
Do EAFP
https://stackoverflow.com/questions/11360858/what-is-the-eafp-principle-in-python

```python
try:
    add_collection_response = add_collection_response.output
except AttributeError:
    pass
```
unless add_collection_response has a possiblity to have "ouput" (which I don't think so)

lmazuel commented on azure-batch/azure/batch/custom/patch.py:
```diff
@@ -0,0 +1,285 @@
+
+        :param collections.deque results_queue: Queue for worker to output results to
+        """
+        while len(self._tasks_to_add) != 0 and not self._has_early_termination_error:
```
```python
while self._tasks_to_add and not self._has_early_termination_error:
```

lmazuel commented on azure-batch/azure/batch/custom/patch.py:
```diff
@@ -0,0 +1,285 @@
+            max_tasks = self._max_tasks_per_request  # local copy
+            chunk_tasks_to_add = []
+            with self._pending_queue_lock:
+                while len(chunk_tasks_to_add) < max_tasks and len(self._tasks_to_add) != 0:
```
```python
while len(chunk_tasks_to_add) < max_tasks and self._tasks_to_add:
```

lmazuel commented on azure-batch/azure/batch/custom/patch.py:
```diff
@@ -0,0 +1,285 @@
+                    self._failures,
+                    self._tasks_to_add)
+            else:
+                return
```
else clause does not serve any purpose here since you're not in a loop

lmazuel commented on azure-batch/azure/batch/custom/patch.py:
```diff
@@ -0,0 +1,285 @@
+        **operation_config)
+
+    # multi-threaded behavior
+    if threads and threads > 0:
```
```python
if threads:
```
In Python, 0 is evaluated as False anyway, so that's equivalent here.

If your problem is <0 values, then test it explicitly and raise a ValueError("thread must be a positive number")

lmazuel commented on azure-batch/azure/batch/custom/patch.py:
```diff
@@ -0,0 +1,285 @@
+    # multi-threaded behavior
+    if threads and threads > 0:
+        active_threads = []
+        for i in range(threads):
```
for i in threads seems to be enough here

lmazuel commented on azure-batch/azure/batch/custom/patch.py:
```diff
@@ -0,0 +1,285 @@
+            active_threads.append(threading.Thread(
+                target=task_workflow_manager.task_collection_thread_handler,
+                args=(results_queue,)))
+            active_threads[i].start()
```
You don't actually need i, you could use -1

bgklein commented:
Addressed most comments minus https://github.com/Azure/azure-sdk-for-python/pull/3217#discussion_r213494559 .  Besides possibly logging traceback at time of occurrence I don't know if there is a uniform way to raise an exception to the user with full traceback

bgklein commented:
@matthchr Could you look at this? Not many changes from extensions.

matthchr commented on azure-batch/azure/batch/batch_service_client.py:
```diff
@@ -23,6 +23,7 @@
 from .operations.task_operations import TaskOperations
 from .operations.compute_node_operations import ComputeNodeOperations
 from . import models
+from .custom.patch import patch_client
```
Do we need to re-add this every single time we regenerate?

I guess there's no other good way...?
Is there a way we can easily know we forgot to do this? I guess the tests will fail...

bgklein replied:
Yea the tests will fail.  They said they are working on adding it to the auto-generation though. So short term tests will fail. Long term it will be done automatically

lmazuel replied:
@matthchr It's in my roadmap to support a "patch.py" file and "custom", it's why I suggested it in the first place.

matthchr commented on azure-batch/azure/batch/custom/patch.py:
```diff
@@ -0,0 +1,287 @@
+            # and resubmit smaller chunk requests
+            # TODO: Replace string with constant variable once available in SDK
+            if e.error.code == "RequestBodyTooLarge":  # pylint: disable=no-member
+                # In this case the task is misbehaved and will not be able to be added due to:
```
This is effectively the same content we had in the extension, right?

matthchr replied:
It looks like it to me

bgklein replied:
Yea only changes were python stylistic stuff Laurent recommended, and a couple adjustments to the objects paths due to being in the python sdk project instead of extensions. 

Only 'new' code is the patch client stuff

bgklein commented:
@annatisch / @lmazuel  Let me know if you have any feedback or if this is good to merge.

lmazuel commented:
@annatisch being your background with batch, I would feel more secure with your approval as well :)

annatisch commented:
Sure thing - I will try to take a look before the end of the week.

annatisch commented on azure-batch/azure/batch/custom/errors.py:
```diff
@@ -0,0 +1,16 @@
+class CreateTasksErrorException(BatchErrorException):
+    """
+    :param str message: Error message describing exit reason
+    :param [~TaskAddResult] failures: List of tasks with detected client side errors.
```
Nevermind - I probably should have started the review from the bottom to get more context :)

annatisch commented on azure-batch/azure/batch/custom/patch.py:
```diff
@@ -0,0 +1,287 @@
+            chunk_tasks_to_add = []
+            with self._pending_queue_lock:
+                while len(chunk_tasks_to_add) < max_tasks and self._tasks_to_add:
+                    chunk_tasks_to_add.append(self._tasks_to_add.pop())
```
I think you should do a try/except IndexError on this - as potentially self._tasks_to_add could have been emptied by another thread in-between the while statement and the pending_queue_lock acquisition.

bgklein replied:
All threads share the same lock + this is the only pop location + conditional self._tasks_to_add should prevent the need?

annatisch replied:
Ah I didn't see the conditional - true, that should handle it.

annatisch commented on azure-batch/azure/batch/custom/patch.py:
```diff
@@ -0,0 +1,287 @@
+                self.error = errors.CreateTasksErrorException(
+                    "One or more tasks failed to be added",
+                    self._failures,
+                    self._tasks_to_add)
```
It seems to me that self._tasks_to_add will only ever have something in it if **_both_** a client error occurred **_and_** an early termination due to an over-sized request body - is that the case?

bgklein replied:
At this line, self._tasks_to_add can be non-empty iff there was an early termination. 
                   self._failures can be non-empty iff there were client-errors to report back to the user.

self._tasks_to_add => contains tasks not yet added

annatisch replied:
Yes - so as I said. I'm not saying it's an issue - just clarifying.

bgklein replied:
You are correct.  Can probably remove from the error contents then because the reason we early terminate is because we want to communicate the original error to the user.

annatisch commented on azure-batch/azure/batch/custom/patch.py:
```diff
@@ -0,0 +1,287 @@
+    else:
+        task_workflow_manager.task_collection_thread_handler(results_queue)
+
+    submitted_tasks = _handle_output(results_queue)
```
I was wondering about moving this line into the else statement below - but I see it can be used to raise any unexpected/unrecognized exceptions that occurred during requests.
I guess it comes down to which exception you wish to take priority in raising to the user - either an error stored in task_workflow_manager.error, and a request exception (if both happened to have occurred).
Your call. :)

bgklein replied:
I feel like if there was an unexpected exception it should be given priority as we would want to explain why we didn't finish serving the request. 
Since we are losing the original stack trace from re-raising it anyways, one idea would be to add it to the custom error raised that way the only exception need to be caught would be CreateTasksErrorException.  Then we could effectively communicate client failures + tasks not submitted + earlyTerminationErrors in one blob..

Thoughts?

bgklein replied:
@matthchr @xingwu1 as well

annatisch replied:
Given that you already have the error structure in place to report back to the user which tasks have/have not been submitted - I think it would make sense to reuse this for other the 'Unrecognized error' scenario... especially something retryable like an intermittent connection error

annatisch commented on azure-batch/tests/test_batch.py:
```diff
@@ -878,6 +878,46 @@ def test_batch_tasks(self, batch_job, **kwargs):
+        result = client.task.add_collection(batch_job.id, tasks_to_add)
+        self.assertIsInstance(result, models.TaskAddCollectionResult)
+        self.assertEqual(len(result.value), 733)
+        self.assertEqual(result.value[0].status, models.TaskAddStatus.success)
```
You could check for the success of all tasks:
`self.assertTrue(all(t.status == models.TaskAddStatus.success for t in result.value))`

bgklein commented on azure-batch/azure/batch/custom/patch.py:
```diff
@@ -0,0 +1,287 @@
+class _TaskWorkflowManager(object):
+    """Worker class for one add_collection request
+
+    :param ExtendedTaskOperations task_operations: Parent object which instantiated this
```
Ugh. Yea its going to be ~TaskOperations

bgklein commented:
@matthchr @xingwu1 Made small changes to error handling so now there is 'new' code.  Feel free to look today if you want. Now it alligns closer with the C# SDK raising an Aggregate Exception of sorts.

@annatisch should be ready for re-review. Let me know if you have any concluding thoughts.

xingwu1 commented:
@lmazuel Could you help to pass the CI? It looks like the jenkins job was stuck.

lmazuel commented:
@xingwu1 This job will test that the package can be built as a wheel (setup.py is correct, etc.). I'm not sure why it's failing, but since I don't see anything that could make me think you changed that, I'm fine to not wait for it.

lmazuel commented:
@bgklein Please remove `azure-batch/__init__.py` from the PR

lmazuel commented on azure-batch/azure/batch/models/__init__.py:
```diff
@@ -9,6 +9,10 @@
 # regenerated.
 # --------------------------------------------------------------------------
 
+######DO NOT REMOVE######
```
You actually don't need to do that, you should be able to inject this from the main `patch` in the client, using importlib to improt "models". I don't guarantee I'm adding this patch point in Autorest.

bgklein commented:
@lmazuel made requested changes

xingwu1 commented:
@lmazuel Could you review and merge it?

bgklein commented:
@lmazuel looks like its still getting stuck at that layer of CI

lmazuel commented:
@bgklein @xingwu1 I'm good to merge and release the PR (I always merge and release at the same time, if you don't want a release, please let me know).

Side note, this means you changed a little of the auto-generated code manually. As we discussed, I'm willing to add support in Autorest for that patch/custom in order to avoid you to manually edit the code each time you have a release. We are adding right now some modifications to Autorest core to allow me to code this (for instance, Autorest is not able today to keep a folder, like we need for "custom"). I don't have ETA on this right now, since I rely on Autorest core team too. In the meantime, please be sure do manually update your code each time you regenerate. Just wanted to be sure we are on the same page on expectations.

Thanks you!

bgklein commented:
@lmazuel Sounds good to us! Thanks for being accommodating!
